lib/autoBlock.js

```diff
@@ -0,0 +1,215 @@
+  }
+
+  if (!config.data) {
+    config.data = {}
+  }
+
+  // Setup noop logging hooks
+  if (!config.onStart) {
+    config.onStart = () => {
+    }
+  }
+
+  if (!config.onSuccess) {
+    config.onSuccess = () => {
+    }
+  }
+
+  if (!config.onFailure) {
+    config.onFailure = () => {
+    }
+  }
+
+  if (!config.onStartStep) {
+    config.onStartStep = () => {
+    }
+  }
+
+  if (!config.onFinishStep) {
+    config.onFinishStep = () => {
+    }
+  }
+
+  config.onStart(config.data)
+
+  // Use optionsMapping and optionsDefaults to build out options if none were explicitly provided on the block
+  if (!config.block.options) {
+    var options = utils.mappingBlock(config.data, config.optionsMapping, config.optionsDefaults)
+    config.block['options'] = {
+      value: options
+    }
+  }
+
+  var autoBlock = buildBlock(config.block, config)
+  var autoCb = (error, results) => finish(error, results, config)
+  try {
+    async.auto(autoBlock, autoCb)
+  } catch (e) {
+    autoCb(e)
+  }
+}
+
+// Resolve the error and results
+function finish (error, results, config) {
+  var data = utils.resultBlock(config, error, results)
+
+  // Map response if desired; can access raw error object
+  if (config.responseMapping) {
+    data.response = utils.mappingBlock(data, config.responseMapping)
+  }
+
+  // Map error if desired; can access response object
+  if (error && !error.requestId && (config.errorMapping || config.errorDefaults)) {
+    error = utils.mapError(data, config.errorMapping, config.errorDefaults)
+  }
+
+  var status = _.get(error, 'data.status')
+  if (error && error.message && status) {
+    error.message = (httpStatus[status] || status) + `: ${error.message}`
+  }
+
+  if (error) {
+    config.onFailure(error, data)
+  } else {
+    config.onSuccess(data.response, data)
+  }
+
+  var done = config.done
+  if (done) {
+    // Lambda will automatically retry if you send the error back
+    if (!_.get(error, 'data.retry', true)) {
+      error = null
+    }
+    process.nextTick(() => done(error, data.response))
+  }
+}
+
+// Build out an async.auto block using our special notation
+function buildBlock (block, config) {
+  var autoBlock = _.transform(block, (asyncBlock, step, key) => {
+    // If no name was provided, use the key. This is used during logging.
+    step.name = step.name || key
+
+    // Figure out which steps need to come before this one
+    // - anything in .after
+    // - the first path segement in each .with
+    // - the first path segement in each .when
+    var deps = (step.after || []).slice()
+
+    var withDeps = (step.with || [])
+    if (_.isString(withDeps)) {
+      withDeps = [withDeps]
+    }
+    withDeps = _.map(withDeps, (value) => _.first(_.split(value, '.')))
+
+    var whenDeps = (step.when || [])
+    if (_.isString(whenDeps)) {
+      whenDeps = [whenDeps]
+    }
+    whenDeps = _.map(whenDeps, (value) => {
+      var dep = _.first(_.split(value, '.'))
+      if (_.first(dep) === '!') {
```

Comment:
It's in the documentation: https://github.com/TrackIF/auto-block/blob/ad317e777dfc19baeec95e2b2612ebc04a87325c/README.md#when

> Negative checks can be made by using a ! prefix:

And it includes example code.

The reason I included this is because it was extremely common to have steps that only ran for simple conditional checks -- but many of them were negations of values.